### ISSUE TITLE
[FR] new features from OneTrainer

### ISSUE THREAD

**Issue by miasik** [LoganBooker/prodigy-plus-schedule-free#38]
OneTrainer has recently got some advanced optimizers and features.
https://github.com/Nerogar/OneTrainer/wiki/Advanced-Optimizers

Is there a chance to see any of the features in PPSF?

**Comment by LoganBooker** [comment]
Hi miasik, PPSF already implements a number of the features those optimiser do (in fact, parts seem to have been inspired by PPSF), including cautious updates, OrthoGrad, factorisation (Adafactor-style, rather than 1-bit), etc.

Probably the main takeaway from those optimisers is the advanced factorisation, but I think it would be problematic for PPSF for a few reasons:

1. Adafactor-style factorisation is already pretty good. While gains could be made by moving to an even more compressed representation, I don't think it's worth the additional maintenance and integration with Prodigy and Schedule-Free's training dynamics (which is very easy to unbalance). 

2. For Schedule-Free, its pretty important the underlying momentum is as precise and high quality as possible, as it is blended with the weights on every step. So any compromise in quality here has a dramatic impact on the optimiser's quality as well. Unlike regular momentum (which you can slim down quite a bit due to how gradient descent works, well documented with stuff like GaLore), applying these technique directly to weights (which is what `z` effectively is) has always failed in the attempts I've made.

So, not ruling anything out, but it's not something I see happening any time soon.

**Comment by miasik** [comment]
> Hi miasik, PPSF already implements a number of the features those optimiser do (in fact, parts seem to have been inspired by PPSF), including cautious updates, OrthoGrad, factorisation (Adafactor-style, rather than 1-bit), etc.

What about Ademamix and Kourkoutas-β?

**Comment by LoganBooker** [comment]
> What about Ademamix and Kourkoutas-β?

Ademamix adds another full momentum state, cancelling out the memory gains from using the factorised second moment. It also complicates both SF and Prodigy, and increases the maintenance burden. I don't have as much time as I used, so I'd prefer not to add anything too significant to the optimiser that would further complicate its inner workings (which are already much heavier than I'd like).

As for Kourkoutas-β, it messes with beta2, and Prodigy applies its own logic to the second momentum updates, so again, it adds another spanner in the wrench. There are already other ways of taming beta2 -- ADOPT, Palm/Adafactor decay, etc. Most of the time, a beta2 of `0.99` works, well, 99% of the time and it's usually not worth stuffing around with unless your own testing shows an issue.

All that said, I am completely open to pull requests that can show these implemented in the optimiser in an elegant, low maintenance way with proven gains, so feel free to have a go!
